packages/picasso.js/src/core/chart/component-collection.js

```diff
@@ -0,0 +1,220 @@
+import createDockLayout from '../layout/dock/docker';
+
+const findComponentByKeyInList = (list, key) => {
+  for (let i = 0; i < list.length; i++) {
+    const currComp = list[i];
+    if (currComp.hasKey && currComp.key === key) {
+      return list[i];
+    }
+  }
+  return null;
+};
+const findComponentByInstanceInList = (list, componentInstance) => {
+  for (let i = 0; i < list.length; i++) {
+    if (list[i].instance === componentInstance) {
+      return list[i];
+    }
+  }
+  return null;
+};
+
+const wrapChildren = (children) =>
+  children?.map((c) => {
+    const dockConfig = c.instance.dockConfig();
+    return {
+      preferredSize(opts) {
+        return dockConfig.computePreferredSize({ ...opts, children: wrapChildren(c.children) });
+      },
+    };
+  });
+const hideAll = (rect, components) => ({ visible: [], hidden: components, order: components });
+const normalLayout = (layoutSettings) => (rect, components) => {
+  const vcomponents = components.map((c) => {
+    const dockConfig = c.instance.dockConfig();
+    return {
+      instance: c.instance,
+      resize: c.instance.resize,
+      preferredSize(opts) {
+        return dockConfig.computePreferredSize({ ...opts, children: wrapChildren(c.children) });
+      },
+      settings: c.settings,
+    };
+  });
+
+  const dockLayout = createDockLayout(layoutSettings);
+
+  const { visible, hidden, order } = dockLayout.layout(rect, vcomponents);
+  return {
+    visible: visible.map((v) => findComponentByInstanceInList(components, v.instance)),
+    hidden: hidden.map((h) => findComponentByInstanceInList(components, h.instance)),
+    order,
+  };
+};
+const customLayout = (fn) => (rect, components) => {
+  const vcomponents = components.map((c, i) => {
+    const dockConfig = c.instance.dockConfig();
+    return {
+      index: i,
+      key: c.settings.key,
+      dockConfig,
+      resize: c.instance.resize,
+      preferredSize(opts) {
+        return dockConfig.computePreferredSize({ ...opts, children: wrapChildren(c.children) });
+      },
+    };
```

Comment:
any suggestion on what information is needed by different types of layout functions?